### DISCUSSION TITLE
Usability - Is it something you are focusing on?

### DISCUSSION THREAD

**Discussion by lurendrejer** [zwave-js/node-red-contrib-zwave-js#143] [Ideas]
I made a long post - nobody likes long posts, so I deleted it.

I've been seeing people in forums using abstraction-layers like home-assistant to get z-wave working in node-red.

In short: Should work be done to intrigue users from the non-technical group or would this be a complete waste of everyones time?
You might already have had these thoughts - but just not chosen to share them anywhere I could find.

The hue, deconz, etc. plugins often has nodes that simply gets triggered by any message, and performs whichever pre-difined (limited) task, for instance: Turn chosen bulb ON.

**Comment by marcus-j-davies** [comment]
Hey,

I have this on my radar, I am in the office currently, so will converse a little later.

**Comment by marcus-j-davies** [comment]
Hey,

Sorry for the delay, I need  a 2nd pair of arms today 😭 

So the node(s) can be seen as more of a framework then anything else, over the past few months we have added some very valuable nodes, that allows very easy integration - with zwave or otherwise devices - I would say some basic level of understanding is needed, but can be something as simple as knowing what a CC is - Command Class (Binary, Multi Level Switch, Indicator) 😉  

Such nodes that are aimed for ease of use

**cmd-factory** is for taking inputs from a message, and constructing a zwave command out of it.

**event-filter** is for splitting the types of data changes out to different pins
i.e a motion sensor may send a battery update, as well as motion data

I have taken the liberty of creating an example flow using the nodes.
The idea: the motion sensor sends out various events - but we only want motion changes, so we filter it.
we create a binary message, and the value for that binary message is actually the motion boolean value.

we use 2 devices (1 set to receive only, another set to send only)

You might need to re-select your nodes - as our numbers might not match,
or the selected devices might not support the involved CCs

To add values to a filter row - activate the filter in the list and double click a value title in the UI - this will give you the option to add it as a filter.

There is nothing stopping you doing some little function nodes to target different types of devices i.e zigbee right after the filter node

![image](https://user-images.githubusercontent.com/55892693/137965312-fd98487c-dd1a-4d37-9e8d-065e898d0998.png)


Below is the export for the flow above.

@crxporter has been the driver behind the more 'user friendly' need for nodes - I just do the hard work 😉 
are we aiming to make it easier  - Indeed! the nodes detailed here take away a surprising amount of complexity.

but a read on the wiki is always recommend   😆 

```
[{"id":"9e95e4896cd43eb4","type":"tab","label":"Flow 2","disabled":false,"info":""},{"id":"2819944a66589548","type":"zwave-device","z":"9e95e4896cd43eb4","name":"Motion Sensor Device","filteredNodeId":"8","multicast":false,"datamode":"Receive","messagesPerMS":1,"messageInterval":250,"isolated":false,"outputs":1,"inputs":0,"x":200,"y":150,"wires":[["4611f7c466f73c54"]]},{"id":"4611f7c466f73c54","type":"event-filter","z":"9e95e4896cd43eb4","name":"Filter Motion Sensor Update","filters":[{"index":0,"name":"Motion Event","valueIds":[{"commandClassName":"Binary Sensor","commandClass":48,"endpoint":0,"property":"Motion","propertyName":"Motion"}],"events":["VALUE_UPDATED"],"strict":false,"id":"355087","_id":0}],"outputs":1,"changeDate":"2021-10-19T17:27:19.657Z","x":250,"y":280,"wires":[["68f5101c1d31ca09"]]},{"id":"68f5101c1d31ca09","type":"cmd-factory","z":"9e95e4896cd43eb4","name":"Trigger Siren (CMD Factory)","node":"","endpoint":"","cc":"Binary Switch","method":"set","params":"[payload.object.newValue]","noEvent":false,"forceUpdate":"","api":"CCAPI","vapiMode":"setValue","vapiValue":"payload","vapiValueId":"valueId","vapiOptions":"options","x":300,"y":485,"wires":[["fde0c40c63f5abbb"]]},{"id":"fde0c40c63f5abbb","type":"zwave-device","z":"9e95e4896cd43eb4","name":"Siren Device","filteredNodeId":"3","multicast":false,"datamode":"Send","messagesPerMS":1,"messageInterval":250,"isolated":false,"outputs":0,"inputs":1,"x":395,"y":645,"wires":[]},{"id":"c04807a65fd9ea19","type":"comment","z":"9e95e4896cd43eb4","name":"Device Node : (In Only Mode)","info":"","x":215,"y":105,"wires":[]},{"id":"ca44a7ef33c289d3","type":"comment","z":"9e95e4896cd43eb4","name":"Only Allow Motion Changes Through","info":"","x":285,"y":235,"wires":[]},{"id":"b529377b574c6c0e","type":"comment","z":"9e95e4896cd43eb4","name":"Construct Binary CMD (using value from Motion Data)","info":"","x":375,"y":410,"wires":[]},{"id":"1d8c4b25f2f9b235","type":"comment","z":"9e95e4896cd43eb4","name":"Send it to siren : (Out Only Mode)","info":"","x":460,"y":605,"wires":[]},{"id":"8ffeb851ca685cad","type":"comment","z":"9e95e4896cd43eb4","name":"We skip many properties, evan node - as our device node next already targets it","info":"","x":455,"y":445,"wires":[]}]
```

**Reply by marcus-j-davies** [reply]
Oh, and node red development being stopped - will never happen (if anything will become more advanced)
NR is used in some very high demand, high availability enterprise applications - and I am a die hard NR fan! 👍

**Reply by lurendrejer** [reply]
Thank you, and thank you for the examples :)
Enterprise applications is the reason I stumbled across node-red and thought "hey, this could work in my home too".

Some projects has templates like yours embedded in the examples tab of node-red - to get people going. I don't really know if it helps anyone - but never the less, it's a thing. I have been fiddling with the cmd-factory and filters for the last few days and i'll be using them to some extend in my work environment.
For now I'll be sticking with my sub-flows to automate my own home:
![image](https://user-images.githubusercontent.com/30922656/138012011-387e7f1b-52f5-4ec7-a51a-85f588078972.png)


And in comparison, this is what a user needs to do, to turn on a zigbee bulb or a z-wave device connected to homeassistant: 
![image](https://user-images.githubusercontent.com/30922656/138011778-5744b74e-cb46-44a9-9cdc-4ecf1a4c4106.png)

And thank you for your time, both here and when guiding me through the CC-jungle.
I enjoy the ccAPI a lot - since it doesn't have the added abstraction that a lot of home automation systems does. Back in the 'Vera' days I would never really know if a parameter or association was saved to the device or just cached in the device-database (whatever a device-database is, anyways).

**Comment by crxporter** [comment]
I guess I'll provide my thoughts since Marcus flagged me! I'm not sure how long you've been watching the project but it has indeed come a *long* way on this route. Less than a year ago it was a single node with difficult-to-build json inputs and difficult-to-understand json outputs. Now we have the ui, souped up device node, event filter, and command factory. 

I'm of course not in charge of where the code goes, but my opinion is this: the purpose of `node-red-contrib-zwave-js` is to provide zwave tools in nodered without catering to any other specific projects.

Things become complicated when someone decides to build for any one external plug-in. Do we pick *my* favorite, nodered homekit like deconz? Do we pick the ominous home assistant or openhab? Do we pick Alexa, google, or wait for Matter? 

I think it's much better to have full access to zwave and provide the tools and documentation for users to integrate that with whatever they like best! I've put in a lot of time lately on the help panels for this node, and the wiki. My dream is to eventually add importable examples using the nodered hamburger menu, to lower the "first day" bar just a bit more... but I don't know when I'll have time to build more on that...

Too long? Didn't read?
My dream for this project is to provide full zwave tools and good documentation about them.

**Reply by lurendrejer** [reply]
Thank you for your answers, and I must clearify - I really respect the scope of the project.
This is exactly what I have always been looking for and I don't want it to change. I can see and feel the passion surrounding the project and i'm pretty sure that everyone like me/us can too.

I discussed the cmd-factory node at work yesterday, with a colleague - on a more 'what is it'-scope.
What is a GUI? -> A GUI is a tool to limit functionality - narrowing the scope.
What is the CMD-factory? -> It's a GUI made by people that finds every part of the functionality important. It has every possible combination of everything + the ability to input raw json, should the user not find everything to be enough.

I teach - and my journey into node-red started from a desire to start teaching IOT and Industrial Automation (2.0). I've had a great amount of fun going from homeseer to node-red and i'll never look back. 

I won't be going anywhere - I was only looking into whether or not the projects vision was clear and from what you just wrote, the scope is very clear.
You don't journey 'into the living room' and ask what people need - Theres no marketing department defining what people needs and you what should care about.

Thank you for taking the time to answer - I'll always be fanfaring zwave-js in home-automation forums.

Renaming can be fun too:
![image](https://user-images.githubusercontent.com/30922656/138010479-f42e1dde-ec46-4a02-8677-618b1d4cd90e.png)

**Reply by lurendrejer** [reply]
> Too long? Didn't read? My dream for this project is to provide full zwave tools and good documentation about them.

Thank you - you have come a long way already, I have never seen any project this comprehensive before - and I love it.

**Comment by lurendrejer** [comment]
My spelling has been complete rubbish throughout these post - I fell out of bed at 02:00 AM and my first thought was to answer your posts on github, I tried my best.

**Reply by marcus-j-davies** [reply]
Well, all I can say is Welcome to the project. Any help, just reach out to us on GitHub or Discord (https://discord.gg/TkybbF7b)
I better go, I am getting emotional 😉

**Comment by marcus-j-davies** [comment]
Hi @lurendrejer, hope your well?

Not sure where you are with this module, you still with us?

Sometime ago, you mentioned packaging example flows with the module that will help new users.
The number of users has increased, and questions around how to use it are coming in.

This is just a Curtesy message to advise we now do so as of the latest version. 
We are also working on 6.6.0 which we will see a new paint job being applied to the ui and mesh maps.
